wpiutil/src/main/native/cpp/RuntimeCheck.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#include "wpi/RuntimeCheck.h"
+
+#ifdef _WIN32
+#include "Windows.h"
+#include <stdio.h>
+#include <memory>
+
+extern "C" int32_t WPI_IsRuntimeValid(uint32_t* foundMajor,
+                                      uint32_t* foundMinor,
+                                      uint32_t* expectedMajor,
+                                      uint32_t* expectedMinor,
+                                      WPI_String* runtimePath) {
+  HMODULE msvcRuntimeModule = GetModuleHandleW(L"msvcp140.dll");
+  if (!msvcRuntimeModule) {
+    return 1;
+  }
+
+  HRSRC versionResource = FindResourceW(msvcRuntimeModule, MAKEINTRESOURCEW(1),
+                                        MAKEINTRESOURCEW(16));
+  if (!versionResource) {
+    return 1;
+  }
+
+  HGLOBAL loadedVersion = LoadResource(msvcRuntimeModule, versionResource);
+  if (!loadedVersion) {
+    return 1;
+  }
+
+  // No need to unlock. Thats vestigial of windows before my time...
+  LPVOID lockedResource = LockResource(loadedVersion);
+  if (!lockedResource) {
+    return 1;
+  }
+
+  DWORD resourceSize = SizeofResource(msvcRuntimeModule, versionResource);
+  if (resourceSize == 0) {
+    return 1;
+  }
+
+  std::unique_ptr<char[]> resourceBuffer =
+      std::make_unique<char[]>(resourceSize);
+  std::memcpy(resourceBuffer.get(), lockedResource, resourceSize);
+
+  VS_FIXEDFILEINFO* fileInfo = nullptr;
+  UINT fileInfoLen = 0;
+
+  if (!VerQueryValueW(resourceBuffer.get(), L"\\",
+                      reinterpret_cast<void**>(&fileInfo), &fileInfoLen)) {
+    return 1;
+  }
+
+  *foundMajor = HIWORD(fileInfo->dwProductVersionMS);
+  *foundMinor = LOWORD(fileInfo->dwProductVersionMS);
+
+  *expectedMajor = 14;
+  *expectedMinor = 40;
```

Comment:
Is there a MSVC define we could use to automatically keep this updated?

Reply:
Not that I could find. The compiler version (_MSC_VER) isn't guaranteed to match the runtime file version. The runtime is only updated every 2 (on average) VS releases.

Reply:
We’d have to do an ifdef chain, but it probably doesn’t matter for us as we always build with latest?  The only potential gain would be if the final else triggered a compiler error, so when the compiler was updated we’d get a reminder to update this value?

Reply:
Yeah, that might be worth doing.

Reply:
Thinking more, thats not worth doing. That would force require a code update every update, which will be hard for future proofing things. Since they don't break this _that_ often, I think this is fine.